Sources/NIO/CircularBuffer.swift

```diff
@@ -205,11 +297,16 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
 
     /// Returns the next index after `index`.
     public func index(after: Int) -> Int {
-        let nextIndex = after + 1
-        precondition(nextIndex <= self.endIndex)
+        let nextIndex = (after + 1) >= self.capacity ? 0 : (after + 1)
         return nextIndex
     }
 
+    /// Returns the index before `index`.
+    public func index(before: Int) -> Int {
+        let previousIndex = (before - 1) < 0 ? self.tailIdx : (before - 1)
```

Comment:
This is mixing indices: it's returning internal indices (`tailIdx`) when it should be returning external ones.

As a separate note, it's also not valid to call `index(before:)` where `before` is `startIndex`, so you can precondition that case away and this degenerates to just a regular subtraction.